hibernate-core/src/main/java/org/hibernate/query/criteria/internal/CriteriaQueryImpl.java

```diff
@@ -258,6 +258,11 @@ public void validate() {
 		if ( getSelection() == null && !hasImplicitSelection() ) {
 			throw new IllegalStateException( "No explicit selection and an implicit one could not be determined" );
 		}
```

Comment:
I believe this check is too strict. SQL 92 says: 

         "A grouped table is a set of groups derived during the evaluation
         of a <group by clause> or a <having clause>. A group is a multiset
         of rows in which all values of the grouping column or columns are
         equal if a <group by clause> is specified, or the group is the
         entire table if no <group by clause> is specified. "

IIUC, HAVING does not require GROUP BY.

Reply:
hmmm, yes... it seems so. In that case I think this part is not good:
```java
if ( ! getGroupings().isEmpty() ) {
	jpaqlQuery.append( " group by " );
	String sep = "";
	for ( Expression grouping : getGroupings() ) {
		jpaqlQuery.append( sep )
				.append( ( (Renderable) grouping ).render( renderingContext ) );
		sep = ", ";
	}

	if ( getHaving() != null ) {
		jpaqlQuery.append( " having " )
				.append( ( (Renderable) getHaving() ).render( renderingContext ) );
	}
}
```
It's from `QueryStructure#render` method. See how it's checking the groupings and only if they are present - adding having clause. Shouldn't that "having" if block be outside the grouping one? WDYT?

Reply:
I agree that doesn't look right.